src/index_map.rs

```diff
@@ -0,0 +1,726 @@
+    #[allow(clippy::iter_not_returning_iterator)]
+    pub fn iter(&self) -> Iter<'_, K, V> {
+        Iter {
+            head_key: &self.head_key,
+            head_val: &self.head_val,
+            iter: std::iter::once((&self.head_key, &self.head_val)).chain(self.tail.iter()),
+        }
+    }
+
+    /// An iterator visiting all elements in their order.
+    #[allow(clippy::iter_not_returning_iterator)]
+    pub fn iter_mut(&mut self) -> IterMut<'_, K, V> {
+        IterMut {
+            iter: std::iter::once((&self.head_key, &mut self.head_val)).chain(self.tail.iter_mut()),
+        }
+    }
+
+    /// An iterator visiting all keys in arbitrary order. The iterator element
+    /// type is `&'a K`.
+    ///
+    /// ```
+    /// use nonempty_collections::*;
+    ///
+    /// let m = ne_indexmap!{"Duke" => "Leto", "Doctor" => "Yueh", "Planetologist" => "Kynes"};
+    /// let v = m.keys().collect::<NEVec<_>>();
+    /// assert_eq!(nev![&"Duke", &"Doctor", &"Planetologist"], v);
+    /// ```
+    pub fn keys(&self) -> Keys<'_, K, V> {
+        Keys {
+            head_key: &self.head_key,
+            inner: std::iter::once(&self.head_key).chain(self.tail.keys()),
+        }
+    }
+
+    /// Returns the number of elements in the map. Always 1 or more.
+    /// ```
+    /// use nonempty_collections::*;
+    ///
+    /// let m = ne_indexmap!{"a" => 1, "b" => 2};
+    /// assert_eq!(2, m.len());
+    /// ```
+    pub fn len(&self) -> usize {
+        self.tail.len() + 1
+    }
+
+    /// A `NEIndexMap` is never empty.
+    #[deprecated(note = "A NEIndexMap is never empty.")]
+    #[allow(clippy::unused_self)]
+    pub const fn is_empty(&self) -> bool {
+        false
+    }
+
+    /// An iterator visiting all values in order.
+    ///
+    /// ```
+    /// use nonempty_collections::*;
+    ///
+    /// let m = ne_indexmap!["Caladan" => "Atreides", "Giedi Prime" => "Harkonnen", "Kaitain" => "Corrino"];
+    /// assert_eq!(vec![&"Atreides", &"Harkonnen", &"Corrino"], m.values().collect::<Vec<_>>());
+    /// ```
+    pub fn values(&self) -> Values<'_, K, V> {
+        Values {
+            head_val: &self.head_val,
+            inner: std::iter::once(&self.head_val).chain(self.tail.values()),
+        }
+    }
+
+    /// Return an iterator visiting all mutable values in order.
+    ///
+    /// ```
+    /// use nonempty_collections::*;
+    /// let mut m = ne_indexmap![0 => "Fremen".to_string(), 1 => "Crysknife".to_string(), 2 => "Water of Life".to_string()];
+    /// m.values_mut().into_iter().for_each(|v| v.truncate(3));
+    /// assert_eq!(vec![&mut "Fre".to_string(), &mut "Cry".to_string(),&mut "Wat".to_string()], m.values_mut().collect::<Vec<_>>());
+    /// ```
+    pub fn values_mut(&mut self) -> ValuesMut<'_, K, V> {
+        ValuesMut {
+            inner: std::iter::once(&mut self.head_val).chain(self.tail.values_mut()),
+        }
+    }
+
+    /// Get the first element. Never fails.
+    pub const fn first(&self) -> (&K, &V) {
```

Comment:
General question: What's the implication of `const` here if `self` is mutated somewhere else?

Reply:
As far as I understand there is none. The `const` keyword signals that it is possible to call this function from a `const` context. [From the language reference](https://doc.rust-lang.org/reference/const_eval.html#const-functions):

> A const fn is a function that one is permitted to call from a const context. Declaring a function const has no effect on any existing uses, it only restricts the types that arguments and the return type may use, as well as prevent various expressions from being used within it. You can freely do anything with a const function that you can do with a regular function.

I personally try to make every function `const` that can be `const`.

Reply:
> The const keyword signals that it is possible to call this function from a const context

Right, it has nothing to do with the value being memoized, etc.